server /api/login endpoint

zship opened the pull request:
This PR implements the `/api/login` endpoint along with:

1. Session management
2. Unit tests and a test runner (because this is the first endpoint implementation). To run tests, use `yarn run tests`.
3. Authn/authz for the `/api/filebrowser` endpoint. This is still a stub on success. It's only meant to demonstrate auth-related failure codes.

A few nonautomated tests you can make against these endpoints are:

```
# run the server
$ yarn run start

# if you use HTTPie like I have in these examples, you'll need a concatenated certificate
$ cat .keys/ca-cert.pem .keys/server-cert.pem > bundle.pem

# a failed login
http --verify=bundle.pem POST https://localhost:8443/api/login username=zach password=hunter3

# login with a missing password param
http --verify=bundle.pem POST https://localhost:8443/api/login username=zach

# a successful login
http --verify=bundle.pem POST https://localhost:8443/api/login username=zach password=hunter2

# test a successfully-authenticated route
$ http --verify=bundle.pem https://localhost:8443/api/filebrowser 'Session-Id:SESSION_ID_FROM_ABOVE_RESPONSE'

# test an invalid session id
$ http --verify=bundle.pem https://localhost:8443/api/filebrowser 'Session-Id:dummy'
```

Users are currently hard-coded into `src/model/user.ts`. If you want to
generate your own, I've provided a script to compute the salt and passwordHash:

```
$ scripts/generate-user
Username: <input your username>
Password: <input your password>
# output:
{
  username: 'zach',
  salt: '7b2c4fb30a242346a359e90f94b043e8',
  passwordHash: 'e9045ae274855d15433b1d5058206ea9def9c4c07041bffd8805d61fc5ab0fa1',
  scopes: [ 'read:filebrowser' ]
}
```

zship committed:
/api/login endpoint. Includes:

1. Session management
2. Unit tests and a test runner (because this is the first endpoint implementation). To run tests, use `yarn run tests`.
3. Authn/authz for the `/api/filebrowser` endpoint. This is still a stub on success. It's only meant to demonstrate auth-related failure codes.

kimlisa commented on server/src/model/session.ts:
```diff
@@ -0,0 +1,80 @@
+    username,
+    scopes,
+    lastActivityTimestamp: now,
+    expirationTimestamp: now + 8 * 60 * 60 * 1000,
```
i saw a max session const somewhere that made it easy to understand it was 8 hours, i would use it here too

zship replied:
Good catch, agreed!

kimlisa commented on server/src/lib/auth.ts:
```diff
@@ -0,0 +1,29 @@
+      password,
+      salt,
+      256 / 8,
+      {
+        N: 2 ** 14,
+        r: 8,
+        p: 4,
+      },
```
i think `N` and `r` are using default values, so `{ p: 4 }` better implies we are using defaults for other values? Also curious why we went with `p: 4`?

zship replied:
Those values were chosen by me throwing a dart at one of [OWASP's recommendations](https://cheatsheetseries.owasp.org/cheatsheets/Password_Storage_Cheat_Sheet.html#scrypt). It would probably make sense for me to add a comment there! `N` and `r` are the default values, yes. That's just me being overly paranoid about those defaults ever changing. I dunno, is that a reasonable fear or not?

alex-kovoy replied:
I would add some comments so these values do not look like magic numbers

kimlisa commented on server/src/route/login/post.ts:
```diff
@@ -0,0 +1,81 @@
+    const err = newError({
+      code: 'login/failed',
+      message: 'password is required',
+    });
```
if i'm following your code correctly, the http code for this will default to `500` since it's not stated, which seems incorrect b/c a missing password in the request is a client side error?

zship replied:
Ah, good catch! `400` is probably the appropriate status code here, will add.

kimlisa commented on server/src/middleware/touchSession.ts:
```diff
@@ -0,0 +1,106 @@
+        type: 'error',
+        error: newError({
+          statusCode: 401,
+          code: 'auth/sessionExpired',
```
should these `code` be consts?

zship replied:
Yes, 100% agree. I would either go with an enum or a type union. Both would avoid typos and enable intellisense completion. My preference has been type unions, I'll get that implemented. If you have a different opinion let me know.

zship replied:
Done: https://github.com/zship/teleport-challenge/blob/ede60a96e604847e106d33c12471a7a2d7951c7b/server/src/lib/ErrorWithCode.ts

kimlisa commented on server/src/model/user.ts:
```diff
@@ -0,0 +1,32 @@
+  if (result.length === 1) {
+    return result[0];
+  }
+  if (result.length > 1) {
+    // would never happen in a database with UNIQUE constraint
+    throw new Error(`Multiple users found matching username ${username}`);
+  }
```
i don't think this throw gets handled by the caller (if it ever did happen)

zship replied:
You're right, it would get propagated through to the last express handler [here](https://github.com/zship/teleport-challenge/blob/9cb91a8587ba5ba4f63d7addc03027998590bbf2/server/src/index.ts#L65) and show to the API client as-is. Which is pretty inelegant, yeah... maybe it *should* be ugly, though, since it should never happen. Hmm.

kimlisa commented on server/src/middleware/touchSession.ts:
```diff
@@ -0,0 +1,106 @@
+      {
+        type: 'error',
+        error: newError({
+          statusCode: 401,
+          code: 'auth/sessionExpired',
```
will there ever be a case where given a custom `code`, the http `statusCode` can be different? if not, i think it would be better to send in the custom `code` only and then assign the appropriate http code

zship replied:
Good point, the way I've written them now is pretty much begging for somebody to eventually edit one and miss  others. Any given `code` should always have the exact same `statusCode`, yes. I was thinking of defining an enumeration of possible `{code, statusCode}` combos, but I think I like your idea better. I could assign `statusCode` in the [jsonErrorHandler](https://github.com/zship/teleport-challenge/blob/9cb91a8587ba5ba4f63d7addc03027998590bbf2/server/src/index.ts#L48). Will do.

zship replied:
Done!

kimlisa commented on server/src/model/session.ts:
```diff
@@ -0,0 +1,80 @@
+    throw newError({
+      statusCode: 401,
+      code: 'auth/sessionInvalid',
+      message: 'The Session-Id used in the request is invalid',
```
i would return a more user friendly message that directs the user to log in?

zship replied:
Done!

alex-kovoy commented on server/src/lib/error.ts:
```diff
@@ -0,0 +1,30 @@
+  return err;
+};
+
+export const isErrorWithCode = (obj: unknown): obj is ErrorWithCode => {
```
I would introduce a new Error class (in Javascript) and use it instead of this parsing code.

zship replied:
Agree, done!

alex-kovoy commented on server/src/middleware/touchSession.ts:
```diff
@@ -0,0 +1,106 @@
+    return;
+  }
+
+  const actions = getActions({
```
I find this logic a bit hard to follow where `getActions` returns a set of actions to execute (?). The caller in this case goes through all actions and decides which one to run. This makes it unclear who is responsible for managing the session state `getActions` method or a `caller`?  What if `getActions` returns 2 actions for a session: `delete` and `update`?

zship replied:
The caller is supposed to unconditionally run all actions returned by `getActions` in the order they are received. `getActions` is the source of truth. If it returns 2 actions `delete` and `update`, that's a bug in `getActions`. My intention is to separate out state mutations (`Session.del()`, `Session.update()`) from the rest of the logic. To make it easier to unit test. Conceptually similar to a state machine definition and interpreter. Does that make sense?

Alternatively I could inject `Session.del()` and `Session.update()` so that unit tests could inject their own implementations. Would that be clearer in your opinion?

alex-kovoy commented on server/src/route/login/post.ts:
```diff
@@ -0,0 +1,81 @@
+> = async (req, res, next) => {
+  const { username, password } = req.body;
+
+  if (username === undefined || password === undefined) {
```
these values may contain an an empty `string` or `null` as well. If so, it's better to use the following:

```
if (!username || !password!)
```

zship replied:
Good catch!

alex-kovoy commented on server/src/route/login/post.ts:
```diff
@@ -0,0 +1,81 @@
+  if (username === undefined || password === undefined) {
+    const err = newError({
+      code: 'login/failed',
+      message: 'username is required',
```
This message is a bit misleading (as input validation message) as it's used for `empty` passwords as well

zship replied:
Yeah, good catch. Late night oversight there. Line 28 should read just `if (username === undefined) {`. Will update.

alex-kovoy commented on server/src/route/login/post.ts:
```diff
@@ -0,0 +1,81 @@
+    return;
+  }
+
+  if (password === undefined) {
```
hm...will this `if` statement ever be executed?

zship replied:
It will not 😳. Same oversight mentioned in [this comment](https://github.com/zship/teleport-challenge/pull/4#discussion_r725680512). Updated!

alex-kovoy commented on server/src/route/login/post.ts:
```diff
@@ -0,0 +1,81 @@
+  // error occurred and handle that error at the very end.
+  let errorDidOccur = false;
+
+  const user: User.User = User.getByUsername(username) || {
```
using an `empty` user to continue the flow can bite as suddenly the flow of handling invalid credentials becomes less intuitive/harder to reason about.

zship replied:
Hmm, I think I see what you mean. Reasoning requires some implicit knowledge (such as the guarantee that [line 63](https://github.com/zship/teleport-challenge/pull/4/files/9cb91a8587ba5ba4f63d7addc03027998590bbf2#diff-bf487af478ff325ba0feae5401d32743c784ab8ea8adda3d664390685a3dbbefR63) will run because an empty `passwordHash` will never match any `getPasswordHash` output). Explicit is pretty much always best, will rewrite.

zship replied:
Okay, an attempt has been made!

zship committed:
PR #4 feedback: use MAX_SESSION_TIME instead of magic number in src/model/session.ts

zship committed:
PR #4 feedback: added comment explaining magic numbers in scrypt() parameters

zship committed:
PR #4 feedback: revamped ErrorWithCode
- Make ErrorWithCode a subclass of Error and use `instanceof` for type
  narrowing rather than an `isErrorWithCode` function.
- Don't assign HTTP status code manually (at ErrorWithCode
  instantiation). Any given `err.code` will always use the same HTTP
  status code, so assign HTTP status code in jsonErrorHandler middleware
  with a mapping (`err.code` -> `httpStatusCode`).

zship committed:
PR #4 feedback: username/password validation should check for both `undefined` and empty strings

zship committed:
PR #4 feedback: make the login flow more explicit about (attempted/best-effort) timing invariance

zship committed:
Overlooked: jsonErrorHandler should serialize ErrorWithCode correctly

zship committed:
PR #4 feedback: made some error message a bit more end-user-friendly
